src/wallet/spend.cpp

```diff
@@ -1392,4 +1392,302 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
+        for (const auto& txIn : tx->vin) {
+            auto prevWalletTx = wallet.GetWalletTx(txIn.prevout.hash);
+            if (prevWalletTx && prevWalletTx->tx->vout[txIn.prevout.n].scriptPubKey == txOut.scriptPubKey) {
+                matchesInput = true;
+                break;
+            }
+        }
+        if (matchesInput) {
+            continue;
+        }
+
+        uniqueTxOutCount++;
+    }
+
+    // we consider two or more unique outputs an "deniabilization" of the coin
+    unsigned int deniabilizationCycles = uniqueTxOutCount >= 2 ? 1 : 0;
+
+    // all txIn and txOut are from our wallet
+    // however if we have multiple txIn this was either an initial deniabilization of multiple UTXOs or the user manually merged deniabilized UTXOs
+    // in either case we don't need to recurse into parent transactions and we can return the calculated cycles
+    if (tx->vin.size() > 1) {
+        return std::make_pair(deniabilizationCycles, false);
+    }
+
+    const auto& txIn = tx->vin[0];
+    // now recursively calculate the deniabilization cycles of the input
+    auto inputStats = CalculateDeniabilizationCycles(wallet, txIn.prevout);
+    return std::make_pair(inputStats.first + deniabilizationCycles, inputStats.second);
+};
+
+util::Result<CreatedTransactionResult> CreateDeniabilizationTransaction(CWallet& wallet, const std::set<COutPoint>& inputs, unsigned int confirm_target, unsigned int deniabilization_cycles, bool sign, bool& insufficient_amount)
+{
+    if (inputs.empty()) {
+        return util::Error{_("Inputs must not be empty")};
+    }
+
+    CCoinControl coin_control = SetupDeniabilizationCoinControl(confirm_target);
+    for (const auto& input : inputs) {
+        coin_control.Select(input);
+    }
+    Assert(coin_control.HasSelected());
+    CFeeRate deniabilization_fee_rate = CalculateDeniabilizationFeeRate(wallet, confirm_target);
+    coin_control.m_feerate = deniabilization_fee_rate;
+
+    LOCK(wallet.cs_wallet);
+
+    FastRandomContext rng_fast;
+    CoinSelectionParams coin_selection_params{rng_fast};
+    auto res_fetch_inputs = FetchSelectedInputs(wallet, coin_control, coin_selection_params);
+    if (!res_fetch_inputs) {
+        return util::Error{util::ErrorString(res_fetch_inputs)};
+    }
+    PreSelectedInputs preset_inputs = *res_fetch_inputs;
+    CAmount total_amount = preset_inputs.total_amount;
+
+    // validate that all UTXOs share the same address
+    std::optional<CTxDestination> op_shared_destination;
+    for (const auto& coin : preset_inputs.coins) {
+        CTxDestination destination = CNoDestination();
+        ExtractDestination(coin->txout.scriptPubKey, destination);
+        if (!std::get_if<CNoDestination>(&destination) && !op_shared_destination) {
+            op_shared_destination = destination;
+        }
+        if (!op_shared_destination || !(*op_shared_destination == destination)) {
+            return util::Error{_("Input addresses must all match.")};
+        }
+    }
```

Comment:
@achow101 - continuing our PubKeyDestination discussion here.
In the highlighted code I was trying to determine if all inputs share the same address or not. 
This seemed to work in all my tests, for real wallets and for unit test wallets (eg from coinbase keys), however with the introduction of the PubKeyDestination address, this now breaks for coinbase keys. In particular the call to OutputTypeFromDestination further down in this function, now returns nullopt.
I suspect my understanding of scripts and addresses is flawed. I assumed scripts is just the general encoding of addresses on the block chain, while addresses (CTxDestination) are specialized classes of address types that Bitcoin Core understands, and I assumed CNoDestination is just addresses we don't understand and can't use for transactions. I'm guessing there's more to it?
How should I modify my code to correctly handle the new PubKeyDestination, which seems to be how coinbase key wallet scripts get classified now?
I added a quick hack to pass the CI below, but it sounds like this is probably not correct behavior (to treat PubKeyDestination as OutputType::Legacy).
Thank you in advance!

Reply:
I suggest that you operate entirely on `CScript`s rather than `CTxDestination`s. Under the previous semantics, P2PK and P2PKH scripts would be bundled together which I don't think is actually the behavior that you want. It would be better to use `CScript` and the `Solver` to get precise script types, and avoid using `CTxDestination` and `OutputType`. The latter two are more for user input and display rather than providing precise distinction between scripts.

Reply:
How do I compare one Script with another (in particular one txout.scriptPubKey with another). 
The idea being that I want to issue a transaction where inputs are strictly from inputs that share the same "address" (address in layman terms - the address that people copy/paste to receive BTC).

Reply:
I think I figured it out. CScript appears to have an operator== so I switched to using that. I'm hoping this works in the way I wanted to compare equality of laymen "addresses".
 
I pushed an update to the PR. Please let me know if this makes sense.